core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java

```diff
@@ -93,7 +93,7 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
             // this is a setting that can only be set within the system!
             IndexMetaData indexMd = allocation.metaData().getIndexSafe(shardRouting.index());
             DiscoveryNodeFilters initialRecoveryFilters = indexMd.getInitialRecoveryFilters();
-            if (shardRouting.recoverySource().getType() != RecoverySource.Type.EXISTING_STORE &&
+            if (RecoverySource.isInitialRecovery(shardRouting.recoverySource().getType()) &&
```

Comment:
just a nit, can we move the `initialRecoveryFilters != null` first since it might be able to skip the lookup then